internal/unionstore/art/art.go

```diff
@@ -0,0 +1,644 @@
+	return t.len
+}
+
+// Size returns the size of the MemBuffer.
+func (t *Art) Size() int {
+	return t.size
+}
+
+// ARTCheckpoint is the checkpoint of memory DB.
+type ARTCheckpoint struct {
+	blockSize     int
+	blocks        int
+	offsetInBlock int
+}
+
+func (cp *ARTCheckpoint) BlockSize() int {
+	return cp.blockSize
+}
+
+func (cp *ARTCheckpoint) Blocks() int {
+	return cp.blocks
+}
+
+func (cp *ARTCheckpoint) OffsetInBlock() int {
+	return cp.offsetInBlock
+}
+
+func (cp *ARTCheckpoint) isSamePosition(other *ARTCheckpoint) bool {
+	return cp.blocks == other.blocks && cp.offsetInBlock == other.offsetInBlock
+}
+
+func (t *Art) checkpoint() ARTCheckpoint {
+	snap := ARTCheckpoint{
+		blockSize: t.allocator.vlogAllocator.blockSize,
+		blocks:    len(t.allocator.vlogAllocator.blocks),
+	}
+	if snap.blocks > 0 {
+		snap.offsetInBlock = t.allocator.vlogAllocator.blocks[snap.blocks-1].length
+	}
+	return snap
+}
+
+// Checkpoint returns a checkpoint of MemDB.
+func (t *Art) Checkpoint() (int, int, int) {
+	cp := t.checkpoint()
+	return cp.blockSize, cp.blocks, cp.offsetInBlock
+}
+
+// RevertToCheckpoint reverts the MemDB to the checkpoint.
+func (t *Art) RevertToCheckpoint(blockSize, blocks, offsetInBlock int) {
+	cp := &ARTCheckpoint{
+		blockSize:     blockSize,
+		blocks:        blocks,
+		offsetInBlock: offsetInBlock,
+	}
+	t.revertToCheckpoint(cp)
+	t.truncate(cp)
+	t.allocator.vlogAllocator.onMemChange()
+}
+
+func (t *Art) Stages() []ARTCheckpoint {
+	return t.stages
+}
+
+func (t *Art) Staging() int {
+	if !t.skipMutex {
+		t.Lock()
+		defer t.Unlock()
+	}
+	t.stages = append(t.stages, t.checkpoint())
+	return len(t.stages)
+}
+
+func (t *Art) Release(h int) {
+	if !t.skipMutex {
+		t.Lock()
+		defer t.Unlock()
+	}
+	if h != len(t.stages) {
+		panic("cannot release staging buffer")
+	}
+	if h == 1 {
+		tail := t.checkpoint()
+		if !t.stages[0].isSamePosition(&tail) {
+			t.dirty = true
+		}
+	}
+	t.stages = t.stages[:h-1]
+}
+
+func (t *Art) Cleanup(h int) {
+	if !t.skipMutex {
+		t.Lock()
+		defer t.Unlock()
+	}
+	if h > len(t.stages) {
+		return
+	}
+	if h < len(t.stages) {
+		panic(fmt.Sprintf("cannot cleanup staging buffer, h=%v, len(db.stages)=%v", h, len(t.stages)))
+	}
+
+	cp := &t.stages[h-1]
+	if !t.vlogInvalid {
+		curr := t.checkpoint()
+		if !curr.isSamePosition(cp) {
+			t.revertToCheckpoint(cp)
+			t.truncate(cp)
+		}
+	}
+	t.stages = t.stages[:h-1]
+	t.allocator.vlogAllocator.onMemChange()
+}
+
+func (t *Art) revertToCheckpoint(cp *ARTCheckpoint) {
+	cursor := t.checkpoint()
+	for !cp.isSamePosition(&cursor) {
+		hdrOff := cursor.offsetInBlock - memdbVlogHdrSize
+		block := t.allocator.vlogAllocator.blocks[cursor.blocks-1].buf
+		var hdr vlogHdr
+		hdr.load(block[hdrOff:])
+		lf := t.allocator.getLeaf(hdr.nodeAddr)
+		lf.vAddr = hdr.oldValue
+		t.size -= int(hdr.valueLen)
+		if hdr.oldValue.isNull() {
+			keptFlags := lf.getKeyFlags()
+			keptFlags = keptFlags.AndPersistent()
+			if keptFlags == 0 {
+				lf.markDelete()
```

Comment:
Why do we only mark it deleted, but not freeing the node as well?

Reply:
Free the node needs to change the parent of it, and I'm not inclined to store the parent address in leaf (which consumes more memory).

Note delete node in the original memdb actually do not release any space, see [comment](https://github.com/tikv/client-go/blob/cd64e24de8d1a4edb9852930fdeddc4abbe07692/internal/unionstore/memdb_arena.go#L282), the only benefit of it is the suppress the height of tree, which is not a issue for ART.

So I think mark it as deleted won't make it worse at least.

Like the [memdb's comment](https://github.com/tikv/client-go/blob/cd64e24de8d1a4edb9852930fdeddc4abbe07692/internal/unionstore/memdb_arena.go#L282), I do not implement reusing leaf not memdb by now, because of the difficulty of length-variety.